python/cudf/cudf/core/dataframe.py

```diff
@@ -2129,36 +2129,6 @@ def _n_largest_or_smallest(self, method, n, columns, keep):
                 df[k] = self[k].reset_index(drop=True).take(new_positions)
         return df.set_index(self.index.take(new_positions))
 
-    def transpose(self):
-        """Transpose index and columns.
-
-        Returns
-        -------
-        a new (ncol x nrow) dataframe. self is (nrow x ncol)
-
-        Notes
-        -----
-        Difference from pandas:
-        Not supporting *copy* because default and only behaviour is copy=True
-        """
-        # Never transpose a MultiIndex - remove the existing columns and
-        # replace with a RangeIndex. Afterward, reassign.
-        columns = self.index.copy(deep=False)
-        index = self.columns.copy(deep=False)
-        if self._num_columns == 0 or self._num_rows == 0:
-            return DataFrame(index=index, columns=columns)
-        # Cython renames the columns to the range [0...ncols]
-        result = self.__class__._from_table(libcudf.transpose.transpose(self))
-        # Set the old column names as the new index
-        result._index = as_index(index)
-        # Set the old index as the new column names
-        result.columns = columns
-        return result
-
-    @property
-    def T(self):
-        return self.transpose()
```

Comment:
@devavret it would be better if these raised a clear `NotImplementedError: do this via cupy instead` or something of the like.

@randerzander @beckernick is it okay if this function just goes away?

Reply:
I'm all on-board the anti-transpose express, but is there any reason we couldn't do it via cupy transparently for them?

Reply:
> I'm all on-board the anti-transpose express, but is there any reason we couldn't do it via cupy transparently for users?

+1 on this as I feel that transpose is a fairly useful general data prep operation for plenty of scenarios in data science.

Reply:
@shwina Since it's impossible to transpose `cudf.Dataframe` => `cudf.Dataframe` when the df contains mixed column types, `df.transpose` is undefined for a lot of use cases. Even transposing same-type columns is slow because it requires random reads and/or random writes. IMO leaving `transpose` on `cudf.Dataframe` is misleading, as cuDF's primary focus is speed/performance. Throwing an error telling users to do `.to_pandas().transpose()` is a more transparent approach.

Reply:
@cwharris Understood - we should totally raise a `NotImplementedError` when the DataFrame contains heterogeneous data types. But if we're going to suggest CuPy as a workaround, we're going to have N different users with N different implementations of that workaround.

Reply:
> we'd see people pulling the dataframe into pandas so that they can still call transpose.

This was the reason we initially added transpose. Because we thought having a slow implementation is better than having users make the round trip through pandas. But even that isn't really an option for users as pandas does a sort of no-data-movement transpose. A transposed pandas dataframe cannot be converted to cuDF dataframe.

Reply:
It's fairly common to convert all your data to be numeric before it goes into the model anyways. Encoding string and categorical features is usually the first part of the data prep process. I would be interested to see what cuML folks think here as well as possibly some of the kaggle folks.

Reply:
> @devavret it would be better if these raised a clear `NotImplementedError: do this via cupy instead` or something of the like.

If we're suggesting cupy, we may also want to demonstrate how to get back to cudf so non-powerusers won't need to look it up elsewhere.